lib/active_merchant/billing/gateways/firstdata_e4_v27.rb

```diff
@@ -0,0 +1,445 @@
+          xml << body
+        end
+
+        xml.target!
+      end
+
+      def build_sale_or_authorization_request(money, credit_card_or_store_authorization, options)
+        xml = Builder::XmlMarkup.new
+
+        add_amount(xml, money, options)
+
+        if credit_card_or_store_authorization.is_a? String
+          add_credit_card_token(xml, credit_card_or_store_authorization, options)
+        else
+          add_credit_card(xml, credit_card_or_store_authorization, options)
+        end
+
+        add_address(xml, options)
+        add_customer_data(xml, options)
+        add_invoice(xml, options)
+        add_tax_fields(xml, options)
+        add_level_3(xml, options)
+
+        xml.target!
+      end
+
+      def build_capture_or_credit_request(money, identification, options)
+        xml = Builder::XmlMarkup.new
+
+        add_identification(xml, identification)
+        add_amount(xml, money, options)
+        add_customer_data(xml, options)
+        add_card_authentication_data(xml, options)
+
+        xml.target!
+      end
+
+      def build_store_request(credit_card, options)
+        xml = Builder::XmlMarkup.new
+
+        add_credit_card(xml, credit_card, options)
+        add_address(xml, options)
+        add_customer_data(xml, options)
+
+        xml.target!
+      end
+
+      def add_credentials(xml)
+        xml.tag! 'ExactID', @options[:login]
+        xml.tag! 'Password', @options[:password]
+      end
+
+      def add_transaction_type(xml, action)
+        xml.tag! 'Transaction_Type', TRANSACTIONS[action]
+      end
+
+      def add_identification(xml, identification)
+        authorization_num, transaction_tag, _ = identification.split(';')
+
+        xml.tag! 'Authorization_Num', authorization_num
+        xml.tag! 'Transaction_Tag', transaction_tag
+      end
+
+      def add_amount(xml, money, options)
+        currency_code = options[:currency] || default_currency
+        xml.tag! 'DollarAmount', localized_amount(money, currency_code)
+        xml.tag! 'Currency', currency_code
+      end
+
+      def add_credit_card(xml, credit_card, options)
+        if credit_card.respond_to?(:track_data) && credit_card.track_data.present?
+          xml.tag! 'Track1', credit_card.track_data
+          xml.tag! 'Ecommerce_Flag', 'R'
+        else
+          xml.tag! 'Card_Number', credit_card.number
+          xml.tag! 'Expiry_Date', expdate(credit_card)
+          xml.tag! 'CardHoldersName', credit_card.name
+          xml.tag! 'CardType', card_type(credit_card.brand)
+
+          add_credit_card_eci(xml, credit_card, options)
+          add_credit_card_verification_strings(xml, credit_card, options)
+        end
+      end
+
+      def add_credit_card_eci(xml, credit_card, options)
+        eci = if credit_card.is_a?(NetworkTokenizationCreditCard) && credit_card.source == :apple_pay && card_brand(credit_card) == 'discover'
+                # Discover requires any Apple Pay transaction, regardless of in-app
```

Comment:
wat.

They should just ignore the ECI then, or automatically set it to `04`. This is silly. (Their requirement, not your code.)

Reply:
Yes. This is a failure of abstraction at many levels (the gateway, FirstData, and the network) that this leaks all the way here, however, doing this is the only way we could have this "just work" for people without knowing this particular piece of arcana. (I'm the one that patched the original FirstData E4 gateway that this is copied from with this logic.)

Reply:
Yeah, I'm _really_ not questioning that the code needs to be here; I just find the gateway requirements a bit silly. It reminds me unpleasantly of when I used to do Win32 coding and you'd end up with `foo.dwSomething = 6; // required for compatibility mode` and the like.